### ISSUE TITLE
live transactions not working 

### ISSUE THREAD

**Issue by GJFox1972** [VivaPayments/API#1256]
Ticket ID CC-1226921

Message from my developer 

Just completed my investigation and it’s 100% issue from the Viva Payment api they have blocked the the Tokenization request for V1 api so it is something they can help with.

I just did a test transaction and it’s working fine so 100% confirm Viva block the request for live account.

Please send them a mail :-
“Live transactions are not working but test are working fine.”

**Comment by matthewrgourd** [comment]
Hi there, sorry to hear about your ongoing issues. If you could first clarify a few things.

By which method are you attempting to integrate?

- [Native Checkout v2 3DS](https://developer.vivawallet.com/online-checkouts/native-checkout-v2/)
- [Redirect Checkout](https://developer.vivawallet.com/online-checkouts/redirect-checkout/)
- [E-commerce plugin](https://developer.vivawallet.com/e-commerce-plugins/) (e.g. [Viva Wallet for WooCommerce](https://developer.vivawallet.com/e-commerce-plugins/woocommerce/), or [Viva Wallet for Shopify](https://developer.vivawallet.com/e-commerce-plugins/shopify/))
- [Payment API integration](https://developer.vivawallet.com/api-reference-guide/payment-api/)

Describe the problem as clearly as possible notifying us of any errors generated.

Please provide any request / response logs generated along with screenshots if possible. Blur out any confidential information.

Also, if you could send through any links to previous related issues that may have been closed prematurely I'd be very grateful. Thanks!

**Comment by GJFox1972** [comment]
Afternoon Matthew,

We are using the API integration 

the integration was tested and it was live 

we then changed the domain from a wordpress staging address to the actual address and now its not processing live transactions 


Regards 
Grant Fox
📞 07789882255
Vulpes Media limited T/A Dreambikecompetition™
reg 11794512

**Comment by GJFox1972** [comment]
Any news here ?

**Comment by matthewrgourd** [comment]
Hi there, sorry, still not enough detail about what it is your developer is attempting to do. Is an e-commerce platform for Wordpress involved, e.g. WooCommerce? Maybe get him to reply directly to us via GitHub here...Thanks! To be honest, 
>Just completed my investigation and it’s 100% issue from the Viva Payment api they have blocked the the Tokenization request for V1 api" 

makes no sense at all.

**Comment by mohitpand** [comment]
Hi there!

When we send the request to this endpoint
"https://api.vivapayments.com/acquiring/v1/cards/chargetokens" to Tokenize
card according to your API help
https://developer.vivawallet.com/api-reference-guide/payment-api/#tag/Cards/paths/~1api~1cards~1chargetokens/post
we receive 403 error which means "You are not authorized to access the
requested resource." according to you API error code details
https://developer.vivawallet.com/api-reference-guide/error-codes/
When I test it with a demo account it's working fine.

Kind regards,
Mohit

**Comment by matthewrgourd** [comment]
Hi @mohitpand the 403 error means there's a problem with authentication. You should be using an OAuth2 bearer token for this? If not, please follow the below instructions before you retry the steps in the tokenization guide:

https://developer.vivawallet.com/authentication/#oauth-2-token-generation

Hope this helps! Let me know how you get on.

**Comment by rikless** [comment]
@matthewrgourd I'm concerned too on this issue. Got 403 on same endpoint.

At this point, I use OAuth2 bearer token and the same request with demo credentials are ok.

We've tried to build the auth bearer and credentials 3 times on 3 different person, and it's still the same response.

What can we do ?

**Comment by matthewrgourd** [comment]
@rikless Are you following the token generation instructions for production when you attempt the switch-over? https://developer.vivawallet.com/authentication/#production-environment

**Comment by rikless** [comment]
@matthewrgourd yes

**Comment by matthewrgourd** [comment]
@rikless please send through any response / request logs so I can check further (blur out any confidential information). If I could also check your account? Email mgourd@vivawallet.com with your merchant ID referring to this GitHub issue. Thanks!

**Comment by rikless** [comment]
@matthewrgourd email send !

**Comment by matthewrgourd** [comment]
Hi @GJFox1972 and @rikless I've spoken to our development team and apparently we've discontinued our Card tokenization API so unfortunately the documentation you've been looking at was incorrect. I've updated the Viva Wallet Developer Portal accordingly. In your calls what you need to do is replace any instances of `/acquiring/v1/cards` with `/nativecheckout/v2` as shown in the below information about Native Checkout v2. The production process for card tokenization should then work. Hope this helps!

**Comment by rikless** [comment]
mmmh, thank you @matthewrgourd but why does it work on sandbox mod ?

**Comment by rikless** [comment]
@matthewrgourd /nativecheckout/v2 endpoint don't allow us to save credit card. What's the issue

**Comment by matthewrgourd** [comment]
@rikless Ok, didn't realise that. I'll get one of our dev team to look at this tomorrow. They're based in Greece so it's already outside office hours now!

**Comment by GJFox1972** [comment]
@mohitpand  gutted!

**Comment by jahazzz** [comment]
Hi @matthewrgourd , I'm currently meeting the same issue as @rikless .
The ability to save cards with the API is one of the most important features of the project I am finishing. I hope you will find a solution quickly ^^

**Comment by matthewrgourd** [comment]
Hi @mohitpand, @rikless, @jahazzz, I spoke to our development team and it looks like the ability to save cards is temporarily unavailable. We'll be reintroducing it soon but until then please refer to the [Native Checkout v2 page](http://developer.vivawallet.com/online-checkouts/native-checkout-v2/) for all the actions that are currently available.

**Comment by rikless** [comment]
@matthewrgourd What do you mean by "temporary" ? What is the solution, because we need to save cards.

**Comment by Vincegypt** [comment]
Hello,

I take over from @rikless to inform you that we are customers at Vivawallet largely for the recurence of payment proposed. We rely on this feature to devellop a new part of our activity. We are therefore looking forward to your return to find out if we need to find another online bank offering recency.

Thank you we are waiting for the reintroduction 

Vincent

**Comment by thefisherbox** [comment]
Hi, 
We have the same problem, we were developing the same functionality but nom it's impossible.
Can You resolve this problem? 
That's really bad.

Regards,
Alexandre

**Comment by rikless** [comment]
@matthewrgourd Can you give us more information please ?

**Comment by Vincegypt** [comment]
Hi everyone, @matthewrgourd if you can give us more concrete information we have spend a lot of money on this and its a big problem for us, please send us some good news.

**Comment by matthewrgourd** [comment]
Hi all, I'm attempting to find out more info from our dev team. Please bear with me.

**Comment by matthewrgourd** [comment]
Hi all, the response from our development team is unfortunately they don't know when the functionality will be restored. We know that many of our customers are disappointed so will try our best to get back to you with a solution.

**Comment by Vincegypt** [comment]
Hi @matthewrgourd , thanks to keep us in touch but we can't accept this answer.... :-)

Please ask them again for a reactivation, we need to finish this project. As i said before we have engaged a lot of money to devellop a plugin with the afirmation from vivawallet that is possible.

We must go forward on this point you have all you need to activate this feature. We can't just wait, some people is working on it we can't stop everything. 
We wait for good news. :-)

thanks 

Vincent

**Comment by rikless** [comment]
@matthewrgourd totally agree with @Vincegypt. We can't accept this answer.

Vivawallet confirmed the availability of this feature before we started to work with our dev team.

Your demo API is still live and never thrown this exception so you need to activate this feature. You've been warned one week ago.

Francois

**Comment by jahazzz** [comment]
@matthewrgourd  We're talking about professionnal application, business and money ...   The least we can expect from a paid service is reactivity and concrete responses with explicit deadlines.
This feature works in sandbox so it can't be so complicated to restore it in production.

**Comment by matthewrgourd** [comment]
I believe we have now also removed it from the sandbox. As I say, we know that many of our customers are disappointed so will try our best to get back to you with a solution.

**Comment by rikless** [comment]
@matthewrgourd We're asking for a specific deadline. Time consumed to remove it from sandbox should have been used to make it live again in production. I really don't understand.

**Comment by Vincegypt** [comment]
Hello @matthewrgourd, 
Really its not possible, its just a reactivation, i'am sure you can understand our point of view.
We have engaged a lot of money because vivawallet is ok with this... 

like @rikless i can't undestand, give us a real date, its not a problem to reactivate this fonction. Please continue to ask. Or better give us the direct contact to solve the situation, our position make us more persuasive, its our money in game....

send us some good news...thanks again

**Comment by matthewrgourd** [comment]
Hi all, sincerest apologies. I've asked the development team again for further info. I'll get back to you if I hear anything.

**Comment by rikless** [comment]
@matthewrgourd I don't understand when you say 

> if I hear anything

Ask your dev team for a response. I think you have to do your best, and can't wait for an optional answer. Handle this issue with your dev team, for real.

**Comment by matthewrgourd** [comment]
@rikless Sure thing, I am chasing to the best of my abilities!

**Comment by Vincegypt** [comment]
Really i'am at the same point as @rikless i don't hear a lot of motivation, it's definitly not your money engaged. Please send us a link to discuss directly with the team in charge of the activations. It's not a problem of abilities, its just not an option for us to activate this, our projects just need this to be finished i don't know if you really understand the situation.

Please add the good person to the conversations or link us directly with them.

Thanks again

**Comment by Vincegypt** [comment]
you can't say something more stressfull than "if i hear something" ....

**Comment by matthewrgourd** [comment]
@Vincegypt, I've just messaged my colleague and asked him to reply on here saying what's going on.

**Comment by Vincegypt** [comment]
Hi @matthewrgourd , ok thanks we wait for more informations.

**Comment by matthewrgourd** [comment]
Hi @Vincegypt, @rikless, @jahazzz, @thefisherbox, @GJFox1972, I've just been informed by our development team that a fix is being worked on for rollout tomorrow. I'll update you again as soon as I have more details.

**Comment by rikless** [comment]
@matthewrgourd Thanks for being in touch. Today we are `tomorrow`. Is that ok ?

**Comment by Vincegypt** [comment]
Hi everyone, yes @matthewrgourd do you have some good news?

**Comment by rikless** [comment]
Hi @matthewrgourd it's now 15 days since you said tomorrow, and we still don't have any news.

**Comment by Vincegypt** [comment]
Hi everyone, yes @matthewrgourd do you have some good news?

**Comment by Vincegypt** [comment]
Hi @matthewrgourd please send us some news, we don't have time for holidays here ... :-) we just want to move forward with this point.

thanks

**Comment by matthewrgourd** [comment]
Hi @rikless, @Vincegypt, I'm finding out for you now (it's my first day back after the Christmas break).

**Comment by Vincegypt** [comment]
Hi @matthewrgourd  thanks

**Comment by matthewrgourd** [comment]
@rikless @Vincegypt apparently it was a bank holiday in Greece yesterday. I didn't know this as I'm based in the UK. Chasing again today...apologies again for the delay.

**Comment by matthewrgourd** [comment]
@GJFox1972, @mohitpand, @rikless, @jahazzz, @Vincegypt, I'm told the functionality is now available again. In fact, apparently it was restored before Christmas I just didn't have the confirmation. Can you try out and let me know if any further issues? Thanks!

**Comment by Vincegypt** [comment]
Hi @matthewrgourd thanks for this good news wa are trying right now ;-)
we keep in touch

**Comment by rikless** [comment]
@matthewrgourd Sorry, but I've tried right now, and this endpoint : 
```https://api.vivapayments.com/acquiring/v1/cards/chargetoken``` still respond as 403 error when using POST.

**Comment by matthewrgourd** [comment]
@rikless, I'm getting the same and have raised with our dev team. Apologies, I'll get back to you as soon as possible.

**Comment by matthewrgourd** [comment]
Hi @rikless, ok, so I'm in the process of updating the documentation but the quick answer is we have replaced the `https://api.vivapayments.com/acquiring/v1/cards/chargetokens` call with `https://api.vivapayments.com/nativecheckout/v2/chargetokens` (`https://demo-api.vivapayments.com/nativecheckout/v2/chargetokens` on demo). Body parameters example:
```
{
  "amount": "1000",
  "cvc": "111",
  "expirationMonth": 1,
  "expirationYear": 2020,
  "holderName": "Card Holder",
  "number": "4111111111111111"
}
```

**Comment by rikless** [comment]
@matthewrgourd we are not charging a card for this request. We want to save the credit card. Are you sure about the body parameters ? (the amount can't be there, and case on property has changed).

**Comment by matthewrgourd** [comment]
@rikless I see what you mean. I'll ask a developer.

**Comment by rikless** [comment]
eventId | 12005

**Comment by matthewrgourd** [comment]
@rikless Awaiting further developer input.

**Comment by matthewrgourd** [comment]
Hi @rikless I'm advised that EventId: 12005 is "ServicesCardTokenizationCardholderAuthenticationFailed". This means that cardholder authentication failed. It is not related to the 403 Forbidden error.

**Comment by matthewrgourd** [comment]
@rikless I'm afraid I gave you the Native Checkout v2 instead of card tokenization endpoints previously hence why the `amount` parameter was mandatory. Sorry for that! Please see below explanation of card tokenization from our dev team:

First you need to generate a charge token. You do this by sending a POST request to https://demo-api.vivapayments.com/nativecheckout/v2/chargetokens

To obtain a card token, make the following GET request including the charge token as a URL parameter:
https://demo-api.vivapayments.com/acquiring/v1/cards/tokens?chargetoken=ctok_Rc4c27orSU56SSMOLDijTMYA8X4

After which a card token will be returned in the response.

Example card token: 05FB1A1EBF41440FDF88A359C46645B6D1EE3EF5

Hope this helps! Please let me know how you get on...

**Comment by rikless** [comment]
We still have: Invalid Amount

**Comment by matthewrgourd** [comment]
@rikless Are you attempting on live or demo?

**Comment by matthewrgourd** [comment]
@rikless Card details need to be valid if testing on live. If testing on demo, card expiration date needs to be in the future. Do not pass an amount until step 4, create transaction.

**Comment by rikless** [comment]
That's what I do.

Another issue there : I see in the mentioned issue that it's not possible to disable 3DS. We store credit card to do offline payment in the future. This doesn't make sens to have 3DS validation for offline payments.

**Comment by Vincegypt** [comment]
Hi everyone, @matthewrgourd whats new, we wait for you and your team, we have no response, its really not acceptable from a service like this.

We wait for this activation for weeks, please give us some real things. we want to move forward

thanks

**Comment by Vincegypt** [comment]
@matthewrgourd  6 days for an answer....its not possible

**Comment by matthewrgourd** [comment]
@Vincegypt, ah, do you still believe there to be a problem? There was a response from me 11 days ago and it was quite detailed. I then closed the issue as I assumed it had been solved. We had switched off card tokenization before Christmas but reinstated it at the beginning of January. This is now how you do the card tokenization in case you missed it the first time I posted!

1. Generate a one-time **charge** token. You do this by sending a POST request to `https://demo-api.vivapayments.com/nativecheckout/v2/chargetokens`.
A one-time charge token will be generated (example: "ctok_Rc4c27orSU56SSMOLDijTMYA8X4")

2. Obtain a **card** token, make the following GET request including the charge token as a URL parameter: `https://demo-api.vivapayments.com/acquiring/v1/cards/tokens?chargetoken=ctok_Rc4c27orSU56SSMOLDijTMYA8X4`.
A unique card token code will be returned in the response (example: "05FB1A1EBF41440FDF88A359C46645B6D1EE3EF5"). 

3. You can make a call to exchange the card token with another one-time charge token as many times as you want using the below call:
GET `https://demo-api.vivapayments.com/acquiring/v1/cards/chargetokens?token=05FB1A1EBF41440FDF88A359C46645B6D1EE3EF5`

4. You can pass the new charge token via the [Create transaction](https://developer.vivawallet.com/online-checkouts/native-checkout-v2/#create-transaction) endpoint without transmitting the real card details of the cardholder again:
POST  `https://demo-api.vivapayments.com/nativecheckout/v2/transactions`

Hope this helps! Please let me know how you get on...

**Comment by Vincegypt** [comment]
@matthewrgourd  i think we still have the same problem @rikless send you a message with no answer and i have call him this afternoon he still have some issues from vivawallet please solve the problem with him
thanks :-)

**Comment by matthewrgourd** [comment]
Hi @Vincegypt the original issue (card tokenization blocked) was fixed two weeks ago. If you or @rikless have something else you want to raise please do so by creating a new ticket. Alternatively, if you're having problems with the offered solution, please elaborate on what these are!

**Comment by Vincegypt** [comment]
Hi @matthewrgourd thanks for your quick answer we really appreciate i contact @rikless asap .

**Comment by rikless** [comment]
@matthewrgourd That's what I do. And amount is still required when we just want to register credit card to make offline payment.

Another issue there : I see in the mentioned issue that it's not possible to disable 3DS. We store credit card to do offline payment in the future. This doesn't make sens to have 3DS validation for offline payments.

**Comment by matthewrgourd** [comment]
@rikless I'm sorry for the delay. Our lead developer is currently away on leave hence the time lag. There were a few things I needed to tidy up in relation to the card tokenization page on our developer portal. The main one is that when you optionally save the card details as a card token, you don't need an amount in the request body: https://developer.vivawallet.com/api-reference-guide/card-tokenization-api/#step-2-generate-card-token-using-the-charge-token-optional . 

Regarding 3DS, when we are talking API level it is not on our side as an acquirer to decide on whether to perform 3DS or not. It depends on a number of things mainly the card, the issuer and the issuer's country. Generally, 3DS authentication will be invoked for all one-off payments. However, if the transaction is recurring, only the first payment is subject to a 3DS prompt.

Hope this helps!

**Comment by Vincegypt** [comment]
@matthewrgourd  i have talk to @rikless , they have test it in demo mode, all is ok and after you have modified the live version and the things go wrong, its possible to ask your team to go back to this fonctionnal version?

**Comment by jahazzz** [comment]
@matthewrgourd you said "when you optionally save the card details as a card token, you don't need an amount in the request body" BUT this is step 2 and in step 1 (charge token request), the amount is mandatory ...

**Comment by matthewrgourd** [comment]
Hi @Vincegypt , if the process flow is working in demo, you can launch to live. The 3DS bit obviously has to be done by a real customer so you won't be able to test this bit until online. As far as I'm aware, there's no offline option.

**Comment by cstsortan** [comment]
Hello, for some reason I'm getting a
`{
    "status": 404,
    "message": "Could not get 'token:charge:<charge_token>', key does not exist",
    "eventId": "2"
}`

when I am trying to use a charge token generated by a card token. Could I be missing some step?

**Comment by matthewrgourd** [comment]
@cstsortan Sounds like the charge token has already been used to take a payment. You need to generate a new one-time charge token for each transaction.

**Comment by rikless** [comment]
@matthewrgourd Hello, will you give any answer to @jahazzz post ? I'm concerned too.

**Comment by matthewrgourd** [comment]
@jahazzz @rikless I'm going to get clarity on this point.

**Comment by matthewrgourd** [comment]
@jahazzz @rikless steps 2 and 3 you don't need anything in the body. This was a mistake in the documentation which I've now rectified. In step 1 `amount` (correctly) is a required parameter. I believe this is because we don't allow anyone to store a customer's card details without first taking a payment for something.

**Comment by rikless** [comment]
@matthewrgourd That's what @jahazzz pointed to you. 

> step 1 (charge token request), the amount is mandatory ...

Please, be careful, that's the issue since https://github.com/VivaPayments/API/issues/1256#issuecomment-573319930

We don't know the amount at this moment, we just want to register a credit card.

**Comment by matthewrgourd** [comment]
@rikless thanks for that. I'm confirming with our dev team why the `payment` parameter is mandatory and if there's any issue there. I think everything else is covered.

**Comment by matthewrgourd** [comment]
Hi @rikless, @jahazzz, @Vincegypt, dev team states the following:

> The amount field is mandatory but any amount that is greater than zero will result in 200 OK and then the card token can be provided without a charge.  It's of very great importance that the cardholder knows his card is being saved in a merchant database and gives his permission for this.

If no further queries I'm going to close this issue.

**Comment by rikless** [comment]
>  It's of very great importance that the cardholder knows his card is being saved in a merchant database and gives his permission for this 

ok but it's not on your end, it's our integration

> The amount field is mandatory but any figure greater than zero will result in 200 OK and then the card token can be provided without a charge

So you are saying : add the mandatory parameter, it will not be used for the charge ?

**Comment by matthewrgourd** [comment]
Hi @rikless yes, I know it's on your end. Sorry if I didn't make that clear. No, the contents of the `amount` parameter in this instance will not be used for the charge.

**Comment by rikless** [comment]
@matthewrgourd when card will ne saved, will we be able to use the card for off session payements, without 3ds ?

Because if it's always required, it's useless to save a credit card
